Extension SDK External API demo (#51)
looker-open-source/extension-examples: react/javascript/demo-external-api/README.md

# Demo External API

This example demonstrates usage of external API in an extension using javascript. Its uses OAUTH2 to get an access token from Google. It then use that token to create and update a spreadsheet using the Google sheet APIs. The key thing to observe are:

- the OauthProvider which handles interfactions with Googles OAUTH2 identity service. The auth token is stored in the pushRouter state so that it survives a page reload.
- The SheetsProvider which provides a simple abstraction to calling the Google sheets restful endpoints.

## Prerequisites

This demo requires a project with access to the sheets API and an OAUTH2 client to be setup in the Google console.

1. [Open the Google console](https://console.cloud.google.com/).
2. Select `IAM and Admin/Create a project`.
3. Enter the name of the project and click `create` which navigates you to the project dashboard.
4. On the project dashboard click `select project` for your newly created project. This displays information about your project and tasks you can do.
5. In the `Getting started` section click `Explore and enable APIs`.

> 6. Click `Libraries` and search for `Sheets`. This brings up a link to the `Google Sheets API`. `Select` it.

Reviewer: ```suggestion
6. Click `Library` and search for `Sheets`. This brings up a link to the `Google Sheets API`. `Select` it.
```

7. Click `Enable` which brings you to the Google Sheets API setup.

> 8. Click the `Configure Consent Screen` button.

Reviewer: This is what I see after `Enable`
![image](https://user-images.githubusercontent.com/6099714/119886466-dc704480-bee7-11eb-9176-945a9aebe285.png)

So I think I need to create the credentials first?

Reviewer: step 1
![image](https://user-images.githubusercontent.com/6099714/119886802-3a9d2780-bee8-11eb-9eae-c7019689e289.png)

Reviewer: After `Done` I get 
![image](https://user-images.githubusercontent.com/6099714/119886986-6b7d5c80-bee8-11eb-8b00-3b9b5a3b93c7.png)

Reviewer: I created some entries there, then I was able to proceed (for the most part the same way) with step 8.

Author: You should not have to do anything with Service accounts

9. Select the `User Type` and click Create. You maybe restricted as to the type you can select.
10. Enter `application information` and click `Save and Continue`. The minimum you need to enter is the `App name`, `User support email` and at least one `developer contact email`.

> 11. Click `Save and Continue` until you see the `Back to Dashboard` button. Click the `Back to Dashboard` button.

Reviewer: I clicked `Save and Continue` twice.

12. Click the `Credentials` link in the sidebar.

> 13. Click the `Create Credentials` button and select `OAUTH client ID`.

Reviewer: ```suggestion
13. Click the `Create Credentials` button and select `OAuth client ID`.
```

14. Select `Application Type` of `Web Application`.
15. `Name` your application.
16. Add the base url of the Looker instance that will host your extension as an `Authorized JavaScript Origin`. Example `https://myinstance.looker.com`.

> 17. Add the base url of the Looker instance that will host your extension appended with `/extensions/oauth2_redirect` as an `Authorized redirect URI`. Example `https://myinstance.looker.com//extensions/oauth2_redirect`.

Reviewer: ```suggestion
17. Add the base url of the Looker instance that will host your extension appended with `/extensions/oauth2_redirect` as an `Authorized redirect URI`. Example `https://myinstance.looker.com/extensions/oauth2_redirect`.
```

Author: Went through the whole process again. I think I have the definitive version now. I added some comments that hopefully give a hint as to how to fix things if things go wrong

  We do this because Looker permissions data access via models. In order to grant / limit access to an extension, it must be associated with a model.

7. Connect the project to Git. This can be done in multiple ways:

- Create a new repository on GitHub or a similar service, and follow the instructions to [connect your project to Git](https://docs.looker.com/data-modeling/getting-started/setting-up-git-connection)
- A simpler but less powerful approach is to set up git with the "Bare" repository option which does not require connecting to an external Git Service.

8. Commit the changes and deploy them to production through the Project UI.

9. Reload the page and click the `Browse` dropdown menu. You will see the extension in the list.

- The extension will load the JavaScript from the `url` provided in the `application` definition. By default, this is http://localhost:8080/bundle.js. If you change the port your server runs on in the package.json, you will need to also update it in the manifest.lkml.
- Refreshing the extension page will bring in any new code changes from the extension template, although some changes will hot reload.

> ## Google Sheets API setup

Reviewer 2: Something that has always slightly bugged me about the format of these READMEs where there are .env files to modify— You must restart your dev server to pick up .env changes (i think).

It's no big deal, but we should either mention that you have to restart the dev server or shift the .env setup to before you `yarn develop`

> The demo requires a project with access to the sheets API and an OAUTH2 client id to be setup in the Google console. To create the project and client id [click here](https://developers.google.com/sheets/api/quickstart/js) and follow the instructions in step 1. The OAUTH client id value needs to be added to a `.env` file in the root directory of this project. This value can be found in the [google developer console](https://console.developers.google.com/). The format of the `.env` file is as follows:

Reviewer 2: I don't think this is the most helpful link / maybe not what you were going for. Maybe this? https://developers.google.com/workspace/guides/create-credentials#web Bit sparse but actually has the credential setup instructions...

> The demo requires a project with access to the sheets API and an OAUTH2 client id to be setup in the Google console. To create the project and client id [click here](https://developers.google.com/sheets/api/quickstart/js) and follow the instructions in step 1. The OAUTH client id value needs to be added to a `.env` file in the root directory of this project. This value can be found in the [google developer console](https://console.developers.google.com/). The format of the `.env` file is as follows:

Reviewer 2: We should be nice and tell them up front to add https://your.looker.com/extensions/oauth2_redirect to the authorized redirect URI list when creating the client ID.

# Demo External API

This example demonstrates usage of external API in an extension using javascript. It uses OAuth to get an access token from Google. It then uses that token to create and update a spreadsheet using the Google sheets APIs. The key things to observe are:

- the OAuthProvider which handles interfactions with Googles OAuth identity service. The auth token is stored in the pushRouter state so that it survives a page reload.
- The SheetsProvider which provides a simple abstraction to calling the Google sheets restful endpoints.

> Note that the demo application uses a sample spreadsheet provided by Google (id `1BxiMVs0XRA5nFMdKvBdBZjgmUUqptlbs74OgvE2upms`). The contents of this spreadsheet is copied to a new spreadsheet that the application can then update.

Reviewer: Good discovery. We should use this magic id for the hackathon app startup too.

# Demo External API

> This example demonstrates usage of external API in an extension using javascript. Its uses OAUTH2 to get an access token from Google. It then use that token to create and update a spreadsheet using the Google sheet APIs. The key thing to observe are:

Reviewer 2: ```suggestion
This example demonstrates usage of external API in an extension using javascript. It uses OAUTH2 to get an access token from Google. It then uses that token to create and update a spreadsheet using the Google sheets APIs. The key things to observe are:
```

Author: Done

8. Click `Save and continue` on the `Edit app registration - Scopes` page.
9. If you selected the `External User type` you will need to enter Test users on the`Edit app registration - Test users` page. Click `Save and Continue`.
10. Click the `Credentials` link in the sidebar.
11. Click the `Create Credentials` button and select `OAuth client ID`.
12. Select `Application Type` of `Web Application`.
13. Enter a `Name` for your application.
14. Add the base url of the Looker instance that will host your extension as an `Authorized JavaScript Origin`. Example `https://myinstance.looker.com`.

> 15. Add the base url of the Looker instance that will host your extension appended with `/extensions/oauth_redirect` as an `Authorized redirect URI`. Example `https://myinstance.looker.com/extensions/oauth_redirect`.

Reviewer: like that url name change

Author: Let me double check that! Thats what I get for doing global change!

16. Click `Create`. This will display the `Client Id` and `Client Secret`. Copy the client id.
17. Create a `.env` file in the root of your project. Create an entry like this: `GOOGLE_CLIENT_ID=Application OAuth client ID` from the previous step. **Important: Do not add the `.env` file to source control.**
18. Click `Library` in the sidebar which will bring up the `Welcome to the API Library Page`.

> 19. Enter `Sheets` in the search filed. This displays a link to the `Google Sheets API`. Click the link.

Reviewer: ```suggestion
19. Enter `Sheets` in the search field. This displays a link to the `Google Sheets API`. Click the link.
```
